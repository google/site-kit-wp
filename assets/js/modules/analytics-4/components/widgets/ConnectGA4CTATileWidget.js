/**
 * ConnectGA4CTATileWidget component.
 *
 * Site Kit by Google, Copyright 2023 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * External dependencies
 */
import PropTypes from 'prop-types';

/**
 * Internal dependencies
 */
import { useSelect } from 'googlesitekit-data';
import ConnectModuleCTATile from '../../../../components/KeyMetrics/ConnectModuleCTATile';
import useWidgetStateEffect from '../../../../googlesitekit/widgets/hooks/useWidgetStateEffect';
import {
	CORE_USER,
	keyMetricsGA4Widgets,
} from '../../../../googlesitekit/datastore/user/constants';
import Null from '../../../../components/Null';

const metadata = { moduleSlug: 'analytics-4' };

export default function ConnectGA4CTATileWidget( { Widget, widgetSlug } ) {
	const ga4DependantKeyMetrics = useSelect( ( select ) => {
		const keyMetrics = select( CORE_USER ).getKeyMetrics();

		if ( ! keyMetrics ) {
			return [];
		}

		return keyMetrics.filter( ( keyMetric ) =>
			keyMetricsGA4Widgets.includes( keyMetric )
		).length;
	} );

	// Only render the CTA if there are 3 or fewer GA4-dependent key metrics tiles.
	// Otherwise, render `Null` to hide the CTA.
	// If there are four GA4-dependent key metrics tiles, the `ConnectGA4CTAWidget` will be rendered instead.
	const ConnectModuleCTAComponent =
		ga4DependantKeyMetrics > 3 ? Null : ConnectModuleCTATile;

	useWidgetStateEffect( widgetSlug, ConnectModuleCTAComponent, metadata );

	// Note that we need to render the `Widget` component as a wrapper here so this component will display
	// correctly when used as a `FallbackComponent` for the `whenActive` HOC. Conversely, when `ConnectModuleCTATile`
	// is rendered as an `OverrideComponent` in `WidgetRenderer` (as a result of multiple adjacent widgets rendering
	// this component and thus sharing the same state), it is wrapped with a `Widget` component - the net
	// result being the ConnectModuleCTATile is correctly wrapped in a `Widget` in both cases.
	return (
		<Widget>
			<ConnectModuleCTATile { ...metadata } />
		</Widget>
	);
}

ConnectGA4CTATileWidget.propTypes = {
	Widget: PropTypes.elementType.isRequired,
	widgetSlug: PropTypes.string.isRequired,
};
