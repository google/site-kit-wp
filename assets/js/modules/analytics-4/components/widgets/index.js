/**
 * GA4 Key Metrics widget components.
 *
 * Site Kit by Google, Copyright 2023 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export { default as ConnectGA4CTATileWidget } from './ConnectGA4CTATileWidget';
export { default as ConnectGA4CTAWidget } from './ConnectGA4CTAWidget';
export { default as EngagedTrafficSourceWidget } from './EngagedTrafficSourceWidget';
export { default as LeastEngagingPagesWidget } from './LeastEngagingPagesWidget';
export { default as MostEngagingPagesWidget } from './MostEngagingPagesWidget';
export { default as NewVisitorsWidget } from './NewVisitorsWidget';
export { default as PopularContentWidget } from './PopularContentWidget';
export { default as PopularProductsWidget } from './PopularProductsWidget';
export { default as ReturningVisitorsWidget } from './ReturningVisitorsWidget';
export { default as TopCitiesWidget } from './TopCitiesWidget';
export { default as TopCitiesDrivingLeadsWidget } from './TopCitiesDrivingLeadsWidget';
export { default as TopCitiesDrivingAddToCartWidget } from './TopCitiesDrivingAddToCartWidget';
export { default as TopCitiesDrivingPurchasesWidget } from './TopCitiesDrivingPurchasesWidget';
export { default as TopDeviceDrivingPurchasesWidget } from './TopDeviceDrivingPurchasesWidget';
export { default as TopCountriesWidget } from './TopCountriesWidget';
export { default as TopTrafficSourceWidget } from './TopTrafficSourceWidget';
export { default as TopTrafficSourceDrivingAddToCartWidget } from './TopTrafficSourceDrivingAddToCartWidget';
export { default as TopTrafficSourceDrivingLeadsWidget } from './TopTrafficSourceDrivingLeadsWidget';
export { default as TopConvertingTrafficSourceWidget } from './TopConvertingTrafficSourceWidget';
export { default as TopTrafficSourceDrivingPurchasesWidget } from './TopTrafficSourceDrivingPurchasesWidget';
export { default as PagesPerVisitWidget } from './PagesPerVisitWidget';
export { default as VisitLengthWidget } from './VisitLengthWidget';
export { default as TopReturningVisitorPages } from './TopReturningVisitorPages';
export { default as VisitsPerVisitorWidget } from './VisitsPerVisitorWidget';
export { default as TopRecentTrendingPagesWidget } from './TopRecentTrendingPagesWidget';
export { default as TopCategoriesWidget } from './TopCategoriesWidget';
export { default as TopPagesDrivingLeadsWidget } from './TopPagesDrivingLeadsWidget';
export { default as PopularAuthorsWidget } from './PopularAuthorsWidget';
