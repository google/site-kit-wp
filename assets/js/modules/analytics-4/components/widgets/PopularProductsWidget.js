/**
 * PopularProductsWidget component.
 *
 * Site Kit by Google, Copyright 2023 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * External dependencies
 */
import PropTypes from 'prop-types';

/**
 * WordPress dependencies
 */
import { compose } from '@wordpress/compose';
import { __ } from '@wordpress/i18n';
import {
	createInterpolateElement,
	useCallback,
	useState,
} from '@wordpress/element';

/**
 * Internal dependencies
 */
import { useSelect, useDispatch, useInViewSelect } from 'googlesitekit-data';
import {
	CORE_USER,
	KM_ANALYTICS_POPULAR_PRODUCTS,
} from '../../../../googlesitekit/datastore/user/constants';
import { CORE_SITE } from '../../../../googlesitekit/datastore/site/constants';
import {
	DATE_RANGE_OFFSET,
	MODULES_ANALYTICS_4,
} from '../../datastore/constants';
import { CORE_UI } from '../../../../googlesitekit/datastore/ui/constants';
import { KEY_METRICS_SELECTION_PANEL_OPENED_KEY } from '../../../../components/KeyMetrics/constants';
import {
	MetricTileTable,
	MetricTileTablePlainText,
} from '../../../../components/KeyMetrics';
import Link from '../../../../components/Link';
import { numFmt } from '../../../../util';
import whenActive from '../../../../util/when-active';
import ConnectGA4CTATileWidget from './ConnectGA4CTATileWidget';
import useViewOnly from '../../../../hooks/useViewOnly';
import withCustomDimensions from '../../utils/withCustomDimensions';

/**
 * Gets the report options for the Popular Products widget.
 *
 * @since 1.127.0
 *
 * @param {Function} select Data store 'select' function.
 * @return {Object} The report options.
 */
function getPopularProductsWidgetReportOptions( select ) {
	const dates = select( CORE_USER ).getDateRangeDates( {
		offsetDays: DATE_RANGE_OFFSET,
	} );

	const productPostType = select( CORE_SITE ).getProductPostType();

	return {
		...dates,
		dimensions: [ 'pagePath' ],
		dimensionFilters: {
			'customEvent:googlesitekit_post_type': {
				filterType: 'stringFilter',
				matchType: 'EXACT',
				value: productPostType,
			},
		},
		metrics: [ { name: 'screenPageViews' } ],
		orderby: [
			{
				metric: { metricName: 'screenPageViews' },
				desc: true,
			},
		],
		limit: 3,
		keepEmptyRows: false,
	};
}

function PopularProductsWidget( props ) {
	const { Widget, WidgetNull } = props;

	const viewOnlyDashboard = useViewOnly();

	const productPostType = useSelect( ( select ) =>
		select( CORE_SITE ).getProductPostType()
	);

	const dates = useSelect( ( select ) =>
		select( CORE_USER ).getDateRangeDates( {
			offsetDays: DATE_RANGE_OFFSET,
		} )
	);

	const { setValue } = useDispatch( CORE_UI );
	const [ showTooltip, setShowTooltip ] = useState( true );

	const openMetricsSelectionPanel = useCallback( () => {
		// Hide the tooltip so it doesn't remain visible, above the panel we're
		// opening.
		setShowTooltip( false );

		// Open the panel.
		setValue( KEY_METRICS_SELECTION_PANEL_OPENED_KEY, true );

		// Wait for the panel to be open before we re-enable the tooltip.
		// This prevents it from appearing above the slide-out panel, see:
		// https://github.com/google/site-kit-wp/issues/7060#issuecomment-1664827831
		setTimeout( () => {
			setShowTooltip( true );
		}, 0 );
	}, [ setValue ] );

	const reportOptions = useSelect( getPopularProductsWidgetReportOptions );

	const isPopularProductsWidgetActive = useSelect( ( select ) =>
		select( CORE_USER ).isKeyMetricActive( KM_ANALYTICS_POPULAR_PRODUCTS )
	);

	const showWidget = isPopularProductsWidgetActive || productPostType;

	const report = useInViewSelect(
		( select ) =>
			showWidget
				? select( MODULES_ANALYTICS_4 ).getReport( reportOptions )
				: undefined,
		[ showWidget, reportOptions ]
	);

	const error = useSelect( ( select ) =>
		select( MODULES_ANALYTICS_4 ).getErrorForSelector( 'getReport', [
			reportOptions,
		] )
	);

	const titles = useInViewSelect(
		( select ) =>
			! error && report
				? select( MODULES_ANALYTICS_4 ).getPageTitles(
						report,
						reportOptions
				  )
				: undefined,
		[ error, report, reportOptions ]
	);

	const loading = useSelect( ( select ) =>
		showWidget
			? ! select( MODULES_ANALYTICS_4 ).hasFinishedResolution(
					'getReport',
					[ reportOptions ]
			  ) || titles === undefined
			: undefined
	);

	const { rows = [] } = report || {};

	const columns = [
		{
			field: 'dimensionValues.0.value',
			Component( { fieldValue } ) {
				const url = fieldValue;
				const title = titles[ url ];
				// Utilizing `useSelect` inside the component rather than
				// returning its direct value to the `columns` array.
				// This pattern ensures that the component re-renders correctly based on changes in state,
				// preventing potential issues with stale or out-of-sync data.
				// Note: This pattern is replicated in a few other spots within our codebase.
				const serviceURL = useSelect( ( select ) => {
					return ! viewOnlyDashboard
						? select( MODULES_ANALYTICS_4 ).getServiceReportURL(
								'all-pages-and-screens',
								{
									filters: {
										unifiedPagePathScreen: url,
									},
									dates,
								}
						  )
						: null;
				} );

				if ( viewOnlyDashboard ) {
					return <MetricTileTablePlainText content={ title } />;
				}

				return (
					<Link
						href={ serviceURL }
						title={ title }
						external
						hideExternalIndicator
					>
						{ title }
					</Link>
				);
			},
		},
		{
			field: 'metricValues.0.value',
			Component( { fieldValue } ) {
				return <strong>{ numFmt( fieldValue ) }</strong>;
			},
		},
	];

	if ( ! showWidget ) {
		return <WidgetNull />;
	}

	const infoTooltip = createInterpolateElement(
		__(
			'Products on your site which visitors viewed the most. Site Kit detected these are your product pages. If this is inaccurate, you can <a>replace</a> this with another metric',
			'google-site-kit'
		),
		{
			a: <Link onClick={ openMetricsSelectionPanel } />,
		}
	);

	let zeroStateMessage = __(
		'Analytics doesn’t have data for your site’s products yet',
		'google-site-kit'
	);

	if ( ! productPostType && isPopularProductsWidgetActive ) {
		zeroStateMessage = __(
			'No product posts currently detected on your site. This metric applies only to sites with product posts.',
			'google-site-kit'
		);
	}

	return (
		<MetricTileTable
			Widget={ Widget }
			widgetSlug={ KM_ANALYTICS_POPULAR_PRODUCTS }
			loading={ loading }
			rows={ rows }
			columns={ columns }
			// Instead of sourcing the infoTooltip from KEY_METRICS_WIDGETS,
			// this widget provides it directly to the MetricTileTable component.
			// This is to accommodate a link behavior within the tooltip when the Metrics Selection Panel is open.
			infoTooltip={ showTooltip ? infoTooltip : null }
			ZeroState={ () => zeroStateMessage }
			error={ error }
			moduleSlug="analytics-4"
		/>
	);
}

PopularProductsWidget.propTypes = {
	Widget: PropTypes.elementType.isRequired,
	WidgetNull: PropTypes.elementType.isRequired,
};

export default compose(
	whenActive( {
		moduleName: 'analytics-4',
		FallbackComponent: ConnectGA4CTATileWidget,
	} ),
	withCustomDimensions( {
		reportOptions: getPopularProductsWidgetReportOptions,
	} )
)( PopularProductsWidget );
