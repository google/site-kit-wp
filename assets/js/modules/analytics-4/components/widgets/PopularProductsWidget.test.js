/**
 * PopularProductsWidget component tests.
 *
 * Site Kit by Google, Copyright 2025 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * External dependencies
 */
import faker from 'faker';
import { capitalize } from 'lodash';

/**
 * Internal dependencies
 */
import { render } from '../../../../../../tests/js/test-utils';
import {
	createTestRegistry,
	provideKeyMetrics,
	provideModules,
	freezeFetch,
	provideUserAuthentication,
	provideSiteInfo,
} from '../../../../../../tests/js/utils';
import { getWidgetComponentProps } from '../../../../googlesitekit/widgets/util';
import {
	CORE_USER,
	KM_ANALYTICS_POPULAR_PRODUCTS,
} from '../../../../googlesitekit/datastore/user/constants';
import PopularProductsWidget from './PopularProductsWidget';
import { withConnected } from '../../../../googlesitekit/modules/datastore/__fixtures__';
import {
	DATE_RANGE_OFFSET,
	MODULES_ANALYTICS_4,
} from '../../datastore/constants';
import { MODULE_SLUG_ANALYTICS_4 } from '../../constants';
import {
	ERROR_INTERNAL_SERVER_ERROR,
	ERROR_REASON_INSUFFICIENT_PERMISSIONS,
} from '../../../../util/errors';
import {
	getAnalytics4MockResponse,
	STRATEGY_ZIP,
} from '../../../analytics-4/utils/data-mock';
import { KEY_METRICS_WIDGETS } from '../../../../components/KeyMetrics/key-metrics-widgets';
import { provideCustomDimensionError } from '../../utils/custom-dimensions';

describe( 'PopularProductsWidget', () => {
	let registry;
	const widgetProps = getWidgetComponentProps(
		KM_ANALYTICS_POPULAR_PRODUCTS
	);
	const reportEndpoint = new RegExp(
		'^/google-site-kit/v1/modules/analytics-4/data/report'
	);
	const accountID = '12345';
	const propertyID = '34567';
	const webDataStreamID = '67890';

	beforeEach( () => {
		registry = createTestRegistry();
		registry.dispatch( CORE_USER ).setReferenceDate( '2020-09-08' );
		provideKeyMetrics( registry );
		provideModules( registry, withConnected( MODULE_SLUG_ANALYTICS_4 ) );
		provideSiteInfo( registry );
		provideUserAuthentication( registry );
		registry.dispatch( MODULES_ANALYTICS_4 ).setSettings( {
			accountID,
			propertyID,
			webDataStreamID,
			availableCustomDimensions:
				KEY_METRICS_WIDGETS[ KM_ANALYTICS_POPULAR_PRODUCTS ]
					.requiredCustomDimensions,
		} );
		registry.dispatch( MODULES_ANALYTICS_4 ).receiveGetProperty(
			{
				createTime: '2014-10-02T15:01:23Z',
			},
			{ propertyID }
		);
		registry
			.dispatch( MODULES_ANALYTICS_4 )
			.receiveIsGatheringData( false );
		registry
			.dispatch( MODULES_ANALYTICS_4 )
			.receiveIsCustomDimensionGatheringData(
				KEY_METRICS_WIDGETS[ KM_ANALYTICS_POPULAR_PRODUCTS ]
					.requiredCustomDimensions[ 0 ],
				false
			);
	} );

	it( 'should render correctly with the expected metrics', async () => {
		const reportOptions = {
			...registry.select( CORE_USER ).getDateRangeDates( {
				offsetDays: DATE_RANGE_OFFSET,
			} ),
			dimensions: [ 'pagePath' ],
			dimensionFilters: {
				'customEvent:googlesitekit_post_type': {
					filterType: 'stringFilter',
					matchType: 'EXACT',
					value: 'product',
				},
			},
			metrics: [ { name: 'screenPageViews' } ],
			orderby: [
				{
					metric: { metricName: 'screenPageViews' },
					desc: true,
				},
			],
			limit: 3,
			keepEmptyRows: false,
		};

		const pageTitlesReportOptions = {
			...registry.select( CORE_USER ).getDateRangeDates( {
				offsetDays: DATE_RANGE_OFFSET,
			} ),
			dimensionFilters: {
				pagePath: new Array( 3 )
					.fill( '' )
					.map( ( _, i ) => `/test-post-${ i + 1 }/` )
					.sort(),
			},
			dimensions: [ 'pagePath', 'pageTitle' ],
			metrics: [ { name: 'screenPageViews' } ],
			orderby: [
				{ metric: { metricName: 'screenPageViews' }, desc: true },
			],
			limit: 15,
		};

		const pageTitlesReport = getAnalytics4MockResponse(
			pageTitlesReportOptions,
			// Use the zip combination strategy to ensure a one-to-one mapping of page paths to page titles.
			// Otherwise, by using the default cartesian product of dimension values, the resulting output will have non-matching
			// page paths to page titles.
			{ dimensionCombinationStrategy: STRATEGY_ZIP }
		);

		pageTitlesReport.rows = pageTitlesReport.rows.map( ( row ) => ( {
			...row,
			dimensionValues: row.dimensionValues.map( ( dimensionValue, i ) =>
				i === 1
					? { value: capitalize( faker.lorem.words( 10 ) ) }
					: dimensionValue
			),
		} ) );

		registry
			.dispatch( MODULES_ANALYTICS_4 )
			.receiveGetReport( pageTitlesReport, {
				options: pageTitlesReportOptions,
			} );

		const report = getAnalytics4MockResponse( reportOptions );

		registry.dispatch( MODULES_ANALYTICS_4 ).receiveGetReport( report, {
			options: reportOptions,
		} );

		const { container, waitForRegistry } = render(
			<PopularProductsWidget { ...widgetProps } />,
			{ registry }
		);
		await waitForRegistry();

		expect( container ).toMatchSnapshot();
	} );

	it( 'should render the loading state while resolving the report', async () => {
		// Freeze the report fetch to keep the widget in loading state.
		freezeFetch( reportEndpoint );

		const { container, waitForRegistry } = render(
			<PopularProductsWidget { ...widgetProps } />,
			{ registry }
		);
		await waitForRegistry();

		[
			'.googlesitekit-km-widget-tile__loading',
			'.googlesitekit-km-widget-tile__loading-header',
			'.googlesitekit-km-widget-tile__loading-body',
		].forEach( ( selector ) => {
			expect( container.querySelector( selector ) ).toBeInTheDocument();
		} );

		expect( container ).toMatchSnapshot();
	} );

	it( 'should render the generic error variant when the report fetch fails', async () => {
		const errorResponse = {
			code: ERROR_INTERNAL_SERVER_ERROR,
			message: 'Internal server error',
			data: { reason: ERROR_INTERNAL_SERVER_ERROR },
		};

		fetchMock.get( reportEndpoint, {
			body: errorResponse,
			status: 500,
		} );

		const { container, getByText, waitForRegistry } = render(
			<PopularProductsWidget { ...widgetProps } />,
			{ registry }
		);

		await waitForRegistry();

		expect( console ).toHaveErrored();

		expect(
			container.querySelector( '.googlesitekit-km-widget-tile--error' )
		).toBeInTheDocument();

		expect( getByText( /Data loading failed/i ) ).toBeInTheDocument();

		expect( container ).toMatchSnapshot();
	} );

	it( 'should render the insufficient permissions error variant when the report fetch fails', async () => {
		const errorResponse = {
			code: 'test_error',
			message: 'Error message.',
			data: { reason: ERROR_REASON_INSUFFICIENT_PERMISSIONS },
		};

		fetchMock.get( reportEndpoint, {
			body: errorResponse,
			status: 500,
		} );

		const { container, getByText, waitForRegistry } = render(
			<PopularProductsWidget { ...widgetProps } />,
			{ registry }
		);

		await waitForRegistry();

		expect( console ).toHaveErrored();

		expect(
			container.querySelector( '.googlesitekit-km-widget-tile--error' )
		).toBeInTheDocument();

		expect( getByText( /Insufficient permissions/i ) ).toBeInTheDocument();

		expect( container ).toMatchSnapshot();
	} );

	it( 'should render the missing custom dimension error when the required custom dimension is not available', async () => {
		registry.dispatch( MODULES_ANALYTICS_4 ).setSettings( {
			accountID,
			propertyID,
			webDataStreamID,
			availableCustomDimensions: [],
		} );

		const { container, getByText, waitForRegistry } = render(
			<PopularProductsWidget { ...widgetProps } />,
			{ registry }
		);

		await waitForRegistry();

		expect(
			container.querySelector( '.googlesitekit-km-widget-tile--error' )
		).toBeInTheDocument();

		expect( getByText( /No data to show/i ) ).toBeInTheDocument();
		expect(
			getByText( /Update Analytics to track metric/i )
		).toBeInTheDocument();

		expect( container ).toMatchSnapshot();
	} );

	it( 'should render the custom dimension error when there is an error creating the custom dimension', async () => {
		const error = {
			code: 'test-error-code',
			message: 'Test error message',
			data: { reason: ERROR_REASON_INSUFFICIENT_PERMISSIONS },
		};

		provideCustomDimensionError( registry, {
			customDimension:
				KEY_METRICS_WIDGETS[ KM_ANALYTICS_POPULAR_PRODUCTS ]
					.requiredCustomDimensions[ 0 ],
			error,
		} );

		const { container, getByText, waitForRegistry } = render(
			<PopularProductsWidget { ...widgetProps } />,
			{ registry }
		);

		await waitForRegistry();

		expect(
			container.querySelector( '.googlesitekit-km-widget-tile--error' )
		).toBeInTheDocument();

		expect( getByText( /Insufficient permissions/i ) ).toBeInTheDocument();

		expect( container ).toMatchSnapshot();
	} );
} );
