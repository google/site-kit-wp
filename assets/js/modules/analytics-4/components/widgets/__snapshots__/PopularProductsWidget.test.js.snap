// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PopularProductsWidget should render correctly with the expected metrics 1`] = `
<div>
  <div
    class="googlesitekit-widget googlesitekit-widget--kmAnalyticsPopularProducts googlesitekit-widget--no-padding"
  >
    <div
      class="googlesitekit-widget__body"
    >
      <div
        class="googlesitekit-km-widget-tile googlesitekit-km-widget-tile--table"
      >
        <div
          class="googlesitekit-km-widget-tile__title-container"
        >
          <h3
            class="googlesitekit-km-widget-tile__title"
          >
            Most popular products by pageviews
          </h3>
          <span
            class="googlesitekit-info-tooltip"
          >
            <svg />
          </span>
        </div>
        <div
          class="googlesitekit-km-widget-tile__body"
        >
          <div
            class="googlesitekit-km-widget-tile__table"
          >
            <div
              class="googlesitekit-table__body-row"
            >
              <div
                class="googlesitekit-table__body-item"
              >
                <a
                  aria-label="Repudiandae beatae praesentium minus ad veritatis sint ut praesentium at (opens in a new tab)"
                  class="googlesitekit-cta-link"
                  href=""
                  rel="noopener noreferrer"
                  target="_blank"
                  title="Repudiandae beatae praesentium minus ad veritatis sint ut praesentium at"
                >
                  <span
                    class="googlesitekit-cta-link__contents"
                  >
                    Repudiandae beatae praesentium minus ad veritatis sint ut praesentium at
                  </span>
                </a>
              </div>
              <div
                class="googlesitekit-table__body-item"
              >
                <strong>
                  100
                </strong>
              </div>
            </div>
            <div
              class="googlesitekit-table__body-row"
            >
              <div
                class="googlesitekit-table__body-item"
              >
                <a
                  aria-label="Illo recusandae et odio quidem rem recusandae in voluptates debitis (opens in a new tab)"
                  class="googlesitekit-cta-link"
                  href=""
                  rel="noopener noreferrer"
                  target="_blank"
                  title="Illo recusandae et odio quidem rem recusandae in voluptates debitis"
                >
                  <span
                    class="googlesitekit-cta-link__contents"
                  >
                    Illo recusandae et odio quidem rem recusandae in voluptates debitis
                  </span>
                </a>
              </div>
              <div
                class="googlesitekit-table__body-item"
              >
                <strong>
                  35
                </strong>
              </div>
            </div>
            <div
              class="googlesitekit-table__body-row"
            >
              <div
                class="googlesitekit-table__body-item"
              >
                <a
                  aria-label="Minima maxime magnam possimus iusto nisi est et aut enim (opens in a new tab)"
                  class="googlesitekit-cta-link"
                  href=""
                  rel="noopener noreferrer"
                  target="_blank"
                  title="Minima maxime magnam possimus iusto nisi est et aut enim"
                >
                  <span
                    class="googlesitekit-cta-link__contents"
                  >
                    Minima maxime magnam possimus iusto nisi est et aut enim
                  </span>
                </a>
              </div>
              <div
                class="googlesitekit-table__body-item"
              >
                <strong>
                  25
                </strong>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`PopularProductsWidget should render the custom dimension error when there is an error creating the custom dimension 1`] = `
<div>
  <div
    class="googlesitekit-km-widget-tile--error"
  >
    <div
      class="googlesitekit-cta googlesitekit-cta--error"
    >
      <div
        class="googlesitekit-cta__header"
      >
        <h2
          class="googlesitekit-cta__header_text"
        >
          Most popular products by pageviews
        </h2>
        <span
          class="googlesitekit-info-tooltip"
          title="Products that brought in the most visitors"
        >
          <svg />
        </span>
      </div>
      <div
        class="googlesitekit-cta__body"
      >
        <h3
          class="googlesitekit-cta__title"
        >
          Insufficient permissions
        </h3>
        
        
        
        
        <div
          class="googlesitekit-report-error-actions"
        >
          <span
            class="googlesitekit-error-retry-text"
          >
            Permissions updated? 
            <button
              class="googlesitekit-cta-link"
            >
              <span
                class="googlesitekit-cta-link__contents"
              >
                Retry
              </span>
            </button>
          </span>
          <span
            class="googlesitekit-error-retry-text"
          >
            Youâ€™ll need to contact your administrator. 
            <a
              aria-label="Learn more (opens in a new tab)"
              class="googlesitekit-cta-link"
              href="https://sitekit.withgoogle.com/support/?error_id=analytics-4_insufficient_permissions"
              rel="noopener noreferrer"
              target="_blank"
            >
              <span
                class="googlesitekit-cta-link__contents"
              >
                Learn more
              </span>
            </a>
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`PopularProductsWidget should render the generic error variant when the report fetch fails 1`] = `
<div>
  <div
    class="googlesitekit-km-widget-tile--error"
  >
    <div
      class="googlesitekit-cta googlesitekit-cta--error"
    >
      <div
        class="googlesitekit-cta__header"
      >
        <h2
          class="googlesitekit-cta__header_text"
        >
          Most popular products by pageviews
        </h2>
        <span
          class="googlesitekit-info-tooltip"
        >
          <svg />
        </span>
      </div>
      <div
        class="googlesitekit-cta__body"
      >
        <h3
          class="googlesitekit-cta__title"
        >
          Data loading failed
        </h3>
        
        
        
        
        <div
          class="googlesitekit-report-error-actions"
        >
          <div
            class="googlesitekit-error-retry-text"
          >
            <a
              aria-label="Get help (opens in a new tab)"
              class="googlesitekit-cta-link"
              href="https://sitekit.withgoogle.com/support/?error_id=internal_server_error"
              rel="noopener noreferrer"
              target="_blank"
            >
              <span
                class="googlesitekit-cta-link__contents"
              >
                Get help
              </span>
            </a>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`PopularProductsWidget should render the insufficient permissions error variant when the report fetch fails 1`] = `
<div>
  <div
    class="googlesitekit-km-widget-tile--error"
  >
    <div
      class="googlesitekit-cta googlesitekit-cta--error"
    >
      <div
        class="googlesitekit-cta__header"
      >
        <h2
          class="googlesitekit-cta__header_text"
        >
          Most popular products by pageviews
        </h2>
        <span
          class="googlesitekit-info-tooltip"
        >
          <svg />
        </span>
      </div>
      <div
        class="googlesitekit-cta__body"
      >
        <h3
          class="googlesitekit-cta__title"
        >
          Insufficient permissions
        </h3>
        
        
        
        
        <div
          class="googlesitekit-report-error-actions"
        >
          <div
            class="googlesitekit-error-retry-text"
          >
            Trouble getting access? 
            <a
              aria-label="Get help (opens in a new tab)"
              class="googlesitekit-cta-link"
              href="https://sitekit.withgoogle.com/support/?error_id=analytics-4_insufficient_permissions"
              rel="noopener noreferrer"
              target="_blank"
            >
              <span
                class="googlesitekit-cta-link__contents"
              >
                Get help
              </span>
            </a>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`PopularProductsWidget should render the loading state while resolving the report 1`] = `
<div>
  <div
    class="googlesitekit-widget googlesitekit-widget--kmAnalyticsPopularProducts googlesitekit-widget--no-padding"
  >
    <div
      class="googlesitekit-widget__body"
    >
      <div
        class="googlesitekit-km-widget-tile googlesitekit-km-widget-tile--table"
      >
        <div
          class="googlesitekit-km-widget-tile__title-container"
        >
          <h3
            class="googlesitekit-km-widget-tile__title"
          >
            Most popular products by pageviews
          </h3>
          <span
            class="screen-reader-text"
          >
            <span
              class="googlesitekit-info-tooltip"
            >
              <svg />
            </span>
          </span>
        </div>
        <div
          class="googlesitekit-km-widget-tile__body"
        >
          <div
            class="googlesitekit-km-widget-tile__loading"
          >
            <div
              class="googlesitekit-preview-block googlesitekit-km-widget-tile__loading-header"
              style="width: 100%; height: 14px;"
            >
              <div
                class="googlesitekit-preview-block__wrapper"
              />
            </div>
            <div
              class="googlesitekit-preview-block googlesitekit-km-widget-tile__loading-body"
              style="width: 100%; height: 53px;"
            >
              <div
                class="googlesitekit-preview-block__wrapper"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`PopularProductsWidget should render the missing custom dimension error when the required custom dimension is not available 1`] = `
<div>
  <div
    class="googlesitekit-km-widget-tile--error"
  >
    <div
      class="googlesitekit-cta googlesitekit-cta--error"
    >
      <div
        class="googlesitekit-cta__header"
      >
        <h2
          class="googlesitekit-cta__header_text"
        >
          Most popular products by pageviews
        </h2>
        <span
          class="googlesitekit-info-tooltip"
          title="Products that brought in the most visitors"
        >
          <svg />
        </span>
      </div>
      <div
        class="googlesitekit-cta__body"
      >
        <h3
          class="googlesitekit-cta__title"
        >
          No data to show
        </h3>
        
        
        
        
        <div
          class="googlesitekit-report-error-actions"
        >
          <button
            class="mdc-button mdc-button--raised"
            target="_self"
          >
            <span
              class="mdc-button__label"
            >
              Update
            </span>
          </button>
          <span
            class="googlesitekit-error-retry-text"
          >
            Update Analytics to track metric
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;
